Add popover previews of (1) conversations with student, (2) response to the background questionnaire, and (3) points

etanttila opened the pull request:
# Description

**What?**

In the views of conversations, add popover previews of:
1. all conversations with student
2. student's response to the background questionnaire (enrollment exercise)
3. student's points (total points for (1) the course overall and by category, (2) the module, and (3) the chapter)
 
Popovers appear on hover and click. If clicked, they are dismissed when clicked anywhere. 
[Screencast from 2024-07-03 17:00:47.webm](https://github.com/apluslms/mooc-jutut/assets/50318434/39ec25a1-0575-4bc5-8bba-64242d3886f4)

The popovers work also on smaller screens and touch screens.
[Screencast from 2024-07-03 17:18:19.webm](https://github.com/apluslms/mooc-jutut/assets/50318434/033bf2e1-36b1-4fbe-801d-3ae9ca017e7e)

The background questionnaire popover displays only the questions that the student responded to and their responses (not all answer options) to make it more compact.

Screenshot of how the points popover appears:
![image](https://github.com/apluslms/mooc-jutut/assets/50318434/df1a74a2-ca4f-432a-acea-453c14aa9388)
(One of the categories is an empty string, which is why there isn't a label below the "A".)
The colors of the progress bar are the same/similar is in A+:
* If there are some exercises in the chapter that have a points-to-pass value and the student hasn't passed some of them, the progress bar for the chapter is displayed red. Similarly for the module progress bar if the student doesn't have enough points to pass the module.
* If the student has full points, the progress bar is displayed green.
* If the student doesn't have full points but enough to pass (or there aren't limitations on points to pass), the progress bar is displayed yellow.

**Why?**

Allow teacher to gain a better understanding of the context (student's background, success on the course, previous conversations) when reading and responding to the student quickly and without having to leave the page.

**How?**

Uses [bootstrap dismiss-on-next-click popovers](https://getbootstrap.com/docs/3.3/javascript/#dismiss-on-next-click).

The contents of a popover are computed the first time it's triggered, to reduce the amount of unnecessary rendering/computations and API calls. Therefore there can be some small delays, especially when displaying the points preview.

For the conversation preview, fetch contents of the conversation page filtering by student, taking just the conversations and removing all unnecessary content (e.g. buttons, feedback tags, etc.) to make it more compact, and insert this into the popover.

For the background questionnaire and points, implement new pages, which are used only to generate the contents for the popovers on demand. The contents are computed by these page's views and then inserted into the popovers.

For the background questionnaires, the enrollment exercise details as well as student responses are cached to reduce API calls and make calculations faster. A new object called BackgroundCache is created to help with setting and getting values to and from the cache.

Fixes #30 (the link part has been fixed before, this now adds the mouseover functionality)
Fixes #31 (doesn't add the direct link, which can be accessed by the student page, but the mouseover functionality, which was preferred by @jsorva)
Fixes #65 


# Testing

**Remember to add or update unit tests for new features and changes.**

* How to [test your changes in A-plus](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations)
* How to [test accessibility](https://wiki.aalto.fi/display/EDIT/How+to+check+the+accessibility+of+pull+requests)


**What type of test did you run?**

- [ ] Accessibility test using the [WAVE](https://wave.webaim.org/extension/) extension.
- [ ] Django unit tests.
- [ ] Selenium tests.
- [ ] Other test. *(Add a description below)*
- [x] Manual testing.

Tested the functionalities work with generated submissions, with different screen widths and also with the phone preview of the inspect tool.
Also tested hat with a multilingual course, the background questionnaire questions and answer options are displayed in the selected language.

**Did you test the changes in**

- [ ] Chrome
- [x] Firefox
- [ ] This pull request cannot be tested in the browser.

**Think of what is affected by these changes and could become broken**

# Translation

- [x] Did you modify or add new strings in the user interface? ([Read about how to create translation](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations))

# Programming style

- [x] Did you follow our [style guides](https://apluslms.github.io/contribute/styleguides/)?
- [x] Did you use Python type hinting in all functions that you added or edited? ([type hints](https://docs.python.org/3/library/typing.html) for function parameters and return values)

# Have you updated the README or other relevant documentation?

- [ ] documents inside the doc directory.
- [ ] README.md.
- [ ] Aplus Manual.
- [x] Other documentation (mention below which documentation).

Tried to add explanatory comments to the code

I have not added documentation to Aplus manual, but indeed that should be added before the end of the summer.

# Is it Done?

- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

*Clean up your git commit history before submitting the pull request!*

etanttila commented:
This change requires a [minor change to the A+ points API](https://github.com/apluslms/a-plus/pull/1380).

ihalaij1 commented on feedback/cached.py:
```diff
@@ -139,3 +148,78 @@ def get_obj(self, key, spec_getter, i18n_getter): # pylint: disable=unused-argum
+        return bgq_dict
+
+    def get_response(self, student: Student, course: Course) -> Optional[tuple[int, dict]]:
+        full_key = '/'.join((
+            self.prefix,
+            self.get_suffix(student.id, course.id, 'response'))
+        )
```
I'd prefer the closing parenthesis of the tuple to be on the same line as the other closing parenthesis (with an extra comma too), like this:
```
full_key = '/'.join((
    self.prefix,
    self.get_suffix(student.id, course.id, 'response'),
))
```

ihalaij1 commented on feedback/cached.py:
```diff
@@ -139,3 +148,78 @@ def get_obj(self, key, spec_getter, i18n_getter): # pylint: disable=unused-argum
+                    student, client, bg_questionnaires
+                )
+            # set value in cache
+            full_key = '/'.join((
+                self.prefix,
+                self.get_suffix(student.id, course.id, 'response'))
+            )
```
I'd prefer the closing parenthesis of the tuple to be on the same line as the other closing parenthesis (with an extra comma too), like this:
```
full_key = '/'.join((
    self.prefix,
    self.get_suffix(student.id, course.id, 'response'),
))
```

ihalaij1 commented on feedback/static/feedback.css:
```diff
@@ -551,6 +558,145 @@ body.ios .feedback-message tr .glyphicon {
+		width: 90vw;
+	}
+}
+.popover:has( .pos-right) {
+	left: auto !important;
+	right: 2rem !important;
+}
```
Is it absolutely necessary to use `!important` here? It is not a good practice to use it in CSS, unless there is no other way to make the styles work properly. If there is no other (relatively easy) way to make this work, add a comment explaining why we use `!important` here. 

https://developer.mozilla.org/en-US/docs/Web/CSS/important#best_practices

etanttila replied:
Good point. The default location is dynamically defined inline, so for the conversation preview it needs to be defined with important, but I'm able to create workarounds for the other popovers.

ihalaij1 commented on feedback/static/feedback.css:
```diff
@@ -551,6 +558,145 @@ body.ios .feedback-message tr .glyphicon {
+	box-shadow: 0 0 0 3px #bce8f1;
+}
+
+/* points display popovers */
+.popover:has( .points-display) {
+	left: auto !important;
+	right: 0 !important;
+	max-width: 400px;
+}
+.popover:has( .points-display) .arrow {
+	left: auto !important;
+	right: 1em !important;
+}
```
Is it absolutely necessary to use `!important` here? It is not a good practice to use it in CSS, unless there is no other way to make the styles work properly. If there is no other (relatively easy) way to make this work, add a comment explaining why we use `!important` here. 

https://developer.mozilla.org/en-US/docs/Web/CSS/important#best_practices

ihalaij1 commented on feedback/views.py:
```diff
@@ -424,7 +439,14 @@ def get_feedback_dict(feedback, get_form, response_form_class, # pylint: disable
 
 
-def update_context_for_feedbacks(request, context, course=None, feedbacks=None, get_form=None, post_url=True): # noqa
+def update_context_for_feedbacks( # noqa
+        request: HttpRequest,
+        context: dict[str, Any],
+        course: Optional[Course] = None,
+        feedbacks: Optional[list[Feedback]] = None,
+        get_form: Optional[Callable[[Feedback], DynamicFeedbacForm]] = None,
+        post_url: bool = True
+        ) -> None:
```
Add a comma after `post_url: bool = True`

ihalaij1 commented on feedback/views.py:
```diff
@@ -483,20 +505,30 @@ def update_context_for_feedbacks(request, context, course=None, feedbacks=None,
+        len(BackgroundCache.get_or_set_bg_questionnaires(course, client)) > 0 # pylint: disable=no-value-for-parameter
+    )
+
+    def student_has_no_bg_questionnaire(student: Student) -> bool:
+        resp = BackgroundCache.get_response(student, course) # pylint: disable=no-value-for-parameter
+        return (resp is not None) and (resp[0] is None)
```
Can this be changed to `student_has_bg_questionnaire`, so that we don't have to use a double negative later on line 571? Currently the line 571 is a bit silly to read:

```python
'show_background': course_has_bg_questionnaire and not student_has_no_bg_questionnaire(conv.student),
```

etanttila replied:
The problem is that this doesn't check if the student has a background questionnaire (as it a student does, but it hasn't been computed yet, it wouldn't appear in the cache). This checks if it has been checked and they student didn't have a background questionnaire (so it doesn't need to be computed and a background questionnaire doesn't need to be displayed). This basically means that the button will show up the first time for a student even if they don't have a background questionnaire, and then if the button is pressed (and the API request is made and we discover the student doesn't have a response to the bg questionnaire, it's saved in the cache so we don't show the button falsely in the future. I guess this could be renamed to "student_might_have_bg_questionnare" or then the entire functionality could be changed so for all students if they don't have a response saved, we actually check if they have a response (and based on that show the button or not), which would slow down the loading somewhat.

ihalaij1 replied:
If its difficult then it doesn't have to be changed. I was just thinking if it was easy, it would be a nice little change.

etanttila replied:
It's not difficult, it just slows down the initial loading time. (For example, if there are about 100 students whose responses haven't been loaded, when testing locally, loading the page with the feedbacks takes about 10 seconds instead of 2.)  But otherwise it does make the UX probably a bit nicer (since the button doesn't appear if the student hasn't filled in the enrollment questionnaire) and indeed does make the code a bit more legible. So it would be a tradeoff. Hard to say which would be better.
I would suspect that decreasing the load time (and fetching the student responses on demand, even if it does mean sometimes in vain showing the "background" button if a student doesn't have a response) would be preferable.

etanttila replied:
Perhaps keeping the functionality as it is and changing the function name to "student_may_have_bg_questionnaire" might be a good way of making the code a bit more legible.

ihalaij1 reviewed (changes requested):
Nice!

I hope I didn't miss any major programming errors, but here are my comments.

I noticed that the videos you linked seem to have empty space under the text of all the buttons, such as the `Respond` button, and the student tags. Compare the `Respond` button and the gray `No response` "tag" to the current one in this picture:
![image](https://github.com/user-attachments/assets/106424ed-226f-4117-bd1e-5b5102b88072)

Do you see how in your video there is empty space, making the buttons and tags longer vertically?

etanttila commented on feedback/cached.py:
```diff
@@ -139,3 +148,78 @@ def get_obj(self, key, spec_getter, i18n_getter): # pylint: disable=unused-argum
+    def get_suffix(self, *args) -> str:
+        return '-'.join(str(x) for x in args)
+
+    def get(self, key: str, course: Course):
+        full_key = '/'.join((self.prefix, self.get_suffix(key, course.id)))
+        return cache.get(full_key)
```
Good point. I thought that since this return value type would be Any, it could be left out. But on further thought, as it is either None or a picklable object (as per [Django documentation](https://docs.djangoproject.com/en/5.0/topics/cache/#django.core.cache.cache.get)), the type could probably be defined as an [`object`](https://docs.djangoproject.com/en/5.0/topics/cache/#django.core.cache.cache.get)

etanttila commented:
> I noticed that the videos you linked seem to have empty space under the text of all the buttons, such as the `Respond` button, and the student tags. Compare the `Respond` button and the gray `No response` "tag" to the current one in this picture: ![image](https://private-user-images.githubusercontent.com/38466145/348152764-106424ed-226f-4117-bd1e-5b5102b88072.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NTMwNDAsIm5iZiI6MTcyMTY1Mjc0MCwicGF0aCI6Ii8zODQ2NjE0NS8zNDgxNTI3NjQtMTA2NDI0ZWQtMjI2Zi00MTE3LWJkMWUtNWI1MTAyYjg4MDcyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIyVDEyNTIyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZlYjM1MjRmMzFiZTM5ODQ5YTg5Yzc3MjgzM2E1ZTY3YWJhMmVlODViZjJjZTUyMTVlNmRkOGU0MDI4MzQzYWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vNdQN_PbVntijqxNhAKnOjMiFFlO1pFVsqMYn8KYrPw)
> 
> Do you see how in your video there is empty space, making the buttons and tags longer vertically?
> 
> I think something in the new CSS is making all the buttons on the page like that.

There indeed seems to be some empty space under the colortags, but that empty space appears also before these commits. (It seems to result from the default padding of the label class from bootstrap, and it hasn't been changed in this commit.)

The empty space below the buttons on the other hand might just be on the video and how that has recorded, it doesn't appear when running the code locally.

ihalaij1 commented:
I was testing the style issues and came across an `AttributeError` in `get_bg_questionnaires`:
![image](https://github.com/user-attachments/assets/fbd4d031-d5ed-4996-b1eb-21f497df76a4)

ihalaij1 commented:
> > I noticed that the videos you linked seem to have empty space under the text of all the buttons, such as the `Respond` button, and the student tags. Compare the `Respond` button and the gray `No response` "tag" to the current one in this picture: ![image](https://private-user-images.githubusercontent.com/38466145/348152764-106424ed-226f-4117-bd1e-5b5102b88072.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NTMwNDAsIm5iZiI6MTcyMTY1Mjc0MCwicGF0aCI6Ii8zODQ2NjE0NS8zNDgxNTI3NjQtMTA2NDI0ZWQtMjI2Zi00MTE3LWJkMWUtNWI1MTAyYjg4MDcyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIyVDEyNTIyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZlYjM1MjRmMzFiZTM5ODQ5YTg5Yzc3MjgzM2E1ZTY3YWJhMmVlODViZjJjZTUyMTVlNmRkOGU0MDI4MzQzYWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vNdQN_PbVntijqxNhAKnOjMiFFlO1pFVsqMYn8KYrPw)
> > Do you see how in your video there is empty space, making the buttons and tags longer vertically?
> > I think something in the new CSS is making all the buttons on the page like that.
> 
> There indeed seems to be some empty space under the colortags, but that empty space appears also before these commits. (It seems to result from the default padding of the label class from bootstrap, and it hasn't been changed in this commit.) (Some additional space may also be from the video, due to low resolution)
> 
> The empty space below the buttons on the other hand might just be on the video and how that has recorded, it doesn't appear when running the code locally.

I tested running it now, it looks like there is no problem with the style now. Might have been just the video quality.

etanttila commented:
> I was testing the style issues and came across an `AttributeError` in `get_bg_questionnaires`. This happened when attempting to go directly to the URL in the screenshot as the `root` user (not accessing through the "Jutut" menu link in A+). ![image](https://private-user-images.githubusercontent.com/38466145/351210581-fbd4d031-d5ed-4996-b1eb-21f497df76a4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE3MzU2MDAsIm5iZiI6MTcyMTczNTMwMCwicGF0aCI6Ii8zODQ2NjE0NS8zNTEyMTA1ODEtZmJkNGQwMzEtZDVlZC00OTk2LWIxZWItMjFmNDk3ZGY3NmE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIzVDExNDgyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdkOTAzMThiN2E2YTM0MTk4ZTE0NzZlZDBkM2FlYWM3OTQ2ZDU0ZDZlNTEyNmJjNWJiMTIzOTM2YzM3NTAyZGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.206Nxw2aqaLR8qvy7RBdStHIIVjegWZt1hRYzAdt57w)

Good observation! I indeed haven't defined fallbacks for if the user doesn't have an aplustoken. I need to add checks in other functions as well that the client is defined, and decide how to deal with cases that there isn't a client. (E.g. should some parts of the page not be displayed, and/or should there be an error message saying that the user should log in via Aplus, or redirecting to an error page.)

etanttila commented:
I now added checks that the client exists (e.g. has A+ api token) and added error messages if it doesn't. Basically these should appear only if the user logs in directly rather than via A+. If relevant information is cached, it's displayed to the user.

![Screenshot from 2024-07-24 14-37-49](https://github.com/user-attachments/assets/0b312979-371d-484d-91fd-7b3d55561b6f)

ihalaij1 reviewed (changes requested):
I found some new issues with the previews that I think weren't there before.

For some reason, the background and points previews here are unnecessarily narrow:
![image](https://github.com/user-attachments/assets/104ee8a2-9748-408a-9019-a4d9d7027e09)

Also, a minor style bug.
When hovering over the `Points` preview, the corner outlines of the button turn from being rounded into square:
![image](https://github.com/user-attachments/assets/ee9f6c6f-fd40-4415-80a2-8ce847121e8b)

This does not happen with the other two buttons:
![image](https://github.com/user-attachments/assets/5c409d8e-10a3-4533-8788-c1530e97ef44)

ihalaij1 reviewed (approved):
I tested again, and this time I flushed my browser cache.
Now the buttons have moved to the right side and the previews are no longer narrow.

Looks good!

ihalaij1 reviewed (changes requested):
Actually, now the conversation preview is really far on the right side?
Is this a new issue?

![image](https://github.com/user-attachments/assets/a2a3782e-3a1b-42e7-9be9-831de30cdf7f)

etanttila commented:
The conversation preview was placed to the right quite intentionally, (so it would cover the conversation as little as possible), but indeed it seems strange that it opens so far away from the button. We discussed with Jimmy that it might be better to move the two buttons ("Show all feedback for this student" and the conversation preview) to the right by the "Background" and "Points" buttons. This way the conversation preview popover could open on the right below the student panel just like the Background popover. This also means that the two buttons don't move around depending on the length of the student's name and the student tags can be immediately after the student name, which would be more intuitive. (The student conversation buttons weren't placed after the student tags because it would lead the location of the buttons to move around even more depending on the number and length of the student tags. Justifying the two buttons to the right resolves that problem.)
Of course, this does mean that we lose a bit of the consistency of having the "Show all feedback for this exercise" right behind the chapter/exercise name and the equivalent button for the student right behind the student name, but fortunately it still stays in the student panel header, and improves the consistency of having the three popover/preview buttons near each other on the right.

So I'll make these changes as discussed with Jimmy.

etanttila commented:
We also noticed when trying with the inspect view's mobile or responsive that when when near the bottom of the page, the page annoyingly scrolls up significantly so that the conversation that was just clicked isn't even visible anymore. Changing the container of the popover to 'body' reduced the issue significantly, even though it didn't fully fix the issue. (I wasn't able to come up with a better solution that would fix it entirely). Jimmy also pointed out that sometimes there may be several popovers open, which was a known issue, caused by manually opening the popovers when they are first triggered (due to the update of the popover content). These don't close unless triggered again, which could be very annoying. Therefore I added a function that manually closes the popover when anything else is clicked or focused after opening the popover (this occurs just once).

I also changed that the background questionnaire preview and and conversations-with-student preview are placed on top for the last conversation on the page if they don't fit below the button.

ihalaij1 reviewed (approved):
Changes look good!
